# Lab | Data Structuring and Combining Data

## Challenge 1: Combining & Cleaning Data

In this challenge, we will be working with the customer data from an insurance company, as we did in the two previous labs. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

But this time, we got new data, which can be found in the following 2 CSV files located at the links below.

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv

Note that you'll need to clean and format the new data.

Observation:
- One option is to first combine the three datasets and then apply the cleaning function to the new combined dataset
- Another option would be to read the clean file you saved in the previous lab, and just clean the two new files and concatenate the three clean datasets

In [123]:
# Your code goes here
import pandas as pd
import numpy as np

df1 = pd.read_csv("file1.csv")
url2 = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv"
df2 = pd.read_csv(url2)
url3 = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv"
df3 = pd.read_csv(url3)

df_list = [df2, df3]

In [124]:
df2.sample(5)

Customer          ST GENDER             Education Customer Lifetime Value  \
199  XO42145      Oregon      M  High School or Below              393381.40%   
847  XA32915      Nevada      M  High School or Below              219973.95%   
704  QF21045      Oregon      F               College              273553.89%   
900  SA40041  Washington      F              Bachelor              574988.38%   
323  MZ24175     Arizona      F               College              309651.12%   

     Income  Monthly Premium Auto Number of Open Complaints  \
199   25927                   103                    1/0/00   
847       0                    64                    1/0/00   
704   99205                    68                    1/0/00   
900   22814                    73                    1/0/00   
323   21604                    79                    1/0/00   

     Total Claim Amount    Policy Type  Vehicle Class  
199          674.973614  Personal Auto            SUV  
847          685.456738  Personal Auto   Two-Door Car  
704          323.246017  Personal Auto   Two-Door Car  
900          360.773462  Personal Auto  Four-Door Car  
323          379.200000  Personal Auto  Four-Door Car

In [125]:

for df in df_list:
    
    columns = df.columns.str.strip().str.replace(' ', '_').str.lower()
    df.columns = columns
    df.rename(columns={df.columns[1]: "state"}, inplace=True)

In [126]:
df3.sample(5)

customer       state  customer_lifetime_value             education  \
6385  LC90366      Nevada             16473.169050              Bachelor   
5012  GQ32356      Nevada              3871.778009  High School or Below   
2795  TO74000      Oregon              8114.409136  High School or Below   
5927  VP84987  California             17497.522010               College   
1082  ZR72236      Nevada              5226.286197                Master   

     gender  income  monthly_premium_auto  number_of_open_complaints  \
6385      M   22828                   106                          0   
5012      F       0                   111                          0   
2795      M   30595                    69                          0   
5927      M       0                    73                          0   
1082      M   76646                    65                          0   

         policy_type  total_claim_amount  vehicle_class  
6385   Personal Auto          508.800000            SUV  
5012   Personal Auto          799.200000  Four-Door Car  
2795  Corporate Auto          446.084760  Four-Door Car  
5927   Personal Auto          350.400000   Two-Door Car  
1082   Personal Auto          263.844471   Two-Door Car

In [127]:
for df in df_list:
    df["gender"] = df["gender"].str.strip().replace({"Femal" : "F", "female" : "F", "Male" : "M"})
    df["state"] = df["state"].str.strip().replace({"Cali" : "California", "AZ" : "Arizona", "WA" : "Washington"})
    df["education"] = df["education"].str.strip().replace("Bachelors", "Bachelor")
    df["customer_lifetime_value"] = df["customer_lifetime_value"].astype(str).str.strip("%").astype(float)
    df["vehicle_class"] = df["vehicle_class"].str.strip().replace({"Luxury Car" : "Luxury", "Luxury SUV" : "Luxury", "Sports Car" : "Luxury"}) 
    df["number_of_open_complaints"] = df["number_of_open_complaints"].astype(str).apply(lambda x: x.split("/")[1] if "/" in x else x)
    df["number_of_open_complaints"] = df["number_of_open_complaints"].replace("nan", np.nan)
    df["number_of_open_complaints"] = df["number_of_open_complaints"].astype(float)
    print(f"{df["number_of_open_complaints"].unique()}")


[0. 1. 3. 5. 2. 4.]
[0. 2. 3. 1. 5. 4.]


In [128]:
df2.isna().sum()

customer                     0
state                        0
gender                       5
education                    0
customer_lifetime_value      4
income                       0
monthly_premium_auto         0
number_of_open_complaints    0
total_claim_amount           0
policy_type                  0
vehicle_class                0
dtype: int64

In [129]:
df3.isna().sum()

customer                     0
state                        0
customer_lifetime_value      0
education                    0
gender                       0
income                       0
monthly_premium_auto         0
number_of_open_complaints    0
policy_type                  0
total_claim_amount           0
vehicle_class                0
dtype: int64

In [130]:

df2['customer_lifetime_value'] = df2['customer_lifetime_value'].fillna(df2['customer_lifetime_value'].mean())
df2['gender'] = df2['gender'].fillna(df2['gender'].mode()[0])
print(df2.isnull().sum())


customer                     0
state                        0
gender                       0
education                    0
customer_lifetime_value      0
income                       0
monthly_premium_auto         0
number_of_open_complaints    0
total_claim_amount           0
policy_type                  0
vehicle_class                0
dtype: int64


In [131]:
df_total = pd.concat([df1, df2, df3])
df_total.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9137 entries, 0 to 7069
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customer                   9137 non-null   object 
 1   state                      9137 non-null   object 
 2   gender                     9137 non-null   object 
 3   education                  9137 non-null   object 
 4   customer_lifetime_value    9137 non-null   float64
 5   income                     9137 non-null   int64  
 6   monthly_premium_auto       9137 non-null   int64  
 7   number_of_open_complaints  9137 non-null   float64
 8   policy_type                9137 non-null   object 
 9   vehicle_class              9137 non-null   object 
 10  total_claim_amount         9137 non-null   float64
dtypes: float64(3), int64(2), object(6)
memory usage: 856.6+ KB


In [132]:
df_total.sample(15)

customer       state gender             education  \
5836  ZX73673     Arizona      M              Bachelor   
3203  ZT52415      Oregon      F              Bachelor   
889   TP87527  California      F  High School or Below   
937   DX80961     Arizona      F              Bachelor   
1193  KZ95954     Arizona      M  High School or Below   
699   YU35386     Arizona      F              Bachelor   
864   DK94262      Oregon      M  High School or Below   
5172  ET44462     Arizona      M  High School or Below   
2028  VZ66567  California      M              Bachelor   
6137  ZX29705  California      F               College   
4255  EO56466      Oregon      M               College   
3463  ET81343  California      M                Doctor   
4999  SI35064  Washington      F              Bachelor   
960   MP76478  Washington      F  High School or Below   
6160  WX52894      Nevada      M               College   

      customer_lifetime_value  income  monthly_premium_auto  \
5836              5555.830205   74015                    72   
3203              8145.985919   62534                    68   
889             239213.110000       0                    65   
937              16265.586820   85322                   134   
1193              5375.657146       0                    80   
699             560001.020000   63621                    70   
864             850712.000000   46754                   106   
5172              5013.136429   65165                   126   
2028              3748.144060       0                   106   
6137              7850.590399       0                    69   
4255             24660.488250   22070                    72   
3463              8750.643581   59047                    76   
4999              9521.676094   74583                   118   
960             864650.410000   64125                   108   
6160             25002.018830   21927                    71   

      number_of_open_complaints     policy_type  vehicle_class  \
5836                        0.0   Personal Auto  Four-Door Car   
3203                        0.0   Personal Auto   Two-Door Car   
889                         0.0  Corporate Auto  Four-Door Car   
937                         0.0   Personal Auto            SUV   
1193                        1.0   Personal Auto  Four-Door Car   
699                         0.0   Personal Auto  Four-Door Car   
864                         1.0   Personal Auto            SUV   
5172                        1.0   Personal Auto            SUV   
2028                        0.0   Personal Auto            SUV   
6137                        0.0  Corporate Auto  Four-Door Car   
4255                        0.0    Special Auto  Four-Door Car   
3463                        3.0   Personal Auto  Four-Door Car   
4999                        0.0   Personal Auto            SUV   
960                         0.0  Corporate Auto            SUV   
6160                        0.0   Personal Auto  Four-Door Car   

      total_claim_amount  
5836          471.443077  
3203          326.400000  
889           361.181122  
937           735.491986  
1193          749.134686  
699           444.620265  
864           513.000000  
5172          604.800000  
2028          508.800000  
6137          331.200000  
4255          345.600000  
3463          364.800000  
4999          161.210468  
960           369.818708  
6160          128.624785

In [133]:
col = df_total.columns
for c in col:
    print(df[c].unique())

['SA25987' 'TB86706' 'ZL73902' ... 'TD14365' 'UP19263' 'Y167826']
['Washington' 'Arizona' 'Nevada' 'California' 'Oregon']
['M' 'F']
['High School or Below' 'Master' 'Bachelor' 'College' 'Doctor']
[3479.137523 2502.637401 3265.156348 ... 8163.890428 7524.442436
 2611.836866]
[    0 25820 30366 ... 47761 71941 21941]
[104  66  82 121 101  70 125  71  64  65  90  62 175 114 108  63  72 137
  80  67 113  88  69  73  61 106 107 132  76  83 103  84 109  74  96 199
 180  81  94 100 119  99 155 122 102  79  68 197 223 129  77 165  97  89
 215  92 126 116 164  91  87  93 105 110 276 117 186 136 172 135 123  86
  98 118 166  95 212 249 134 128 181 139 187 111 240 196  78 189 115  85
 225 124 127 130 192 112 201 245 193 131 295 191 239 195 216 159 173 182
 194 138 133 149 185 144 211 214 179 232 244 146 183 154 198 143 140 190
 158 207 283 184 238 153 229 220 178 210 204 152 213 177 281 170 237 151
 234 171 188 161 218 275 209 148 261 157 167 205 141 228 242 160 253 284
 256 231 208 296 235 285 1

# Challenge 2: Structuring Data

In this challenge, we will continue to work with customer data from an insurance company, but we will use a dataset with more columns, called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by performing data cleaning, formatting, and structuring.

In [ ]:
# Your code goes here
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv"
df = pd.read_csv(url)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 27 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   unnamed:_0                     10910 non-null  int64  
 1   customer                       10910 non-null  object 
 2   state                          10910 non-null  object 
 3   customer_lifetime_value        10910 non-null  float64
 4   response                       10910 non-null  object 
 5   coverage                       10910 non-null  object 
 6   education                      10910 non-null  object 
 7   effective_to_date              10910 non-null  object 
 8   employmentstatus               10910 non-null  object 
 9   gender                         10910 non-null  object 
 10  income                         10910 non-null  int64  
 11  location_code                  10910 non-null  object 
 12  marital_status                 10910 non-null 

In [135]:
df.sample(10)

unnamed:_0 customer       state  customer_lifetime_value response  \
4292         4292  KL58949  California              5577.942343       No   
4356         4356  AP46166  California              4965.491316       No   
1511         1511  HP42153  Washington              2351.878675       No   
1571         1571  FL50532     Arizona              8237.251407       No   
7068         7068  UT38865      Oregon              7425.870642       No   
9230         9230  KZ60207     Arizona              5007.672425       No   
7871         7871  AW84946     Arizona              4479.598820       No   
4431         4431  KS12490      Oregon             10594.132490       No   
9252         9252  IG80910      Oregon              3071.784639       No   
10509       10509  TZ65661  California              2742.797845       No   

       coverage             education effective_to_date employmentstatus  \
4292      Basic              Bachelor        2011-02-18         Employed   
4356      Basic  High School or Below        2011-01-06       Unemployed   
1511      Basic  High School or Below        2011-01-29    Medical Leave   
1571      Basic                Master        2011-02-28         Employed   
7068      Basic               College        2011-01-07         Employed   
9230      Basic               College        2011-01-16         Employed   
7871      Basic              Bachelor        2011-02-04         Employed   
4431   Extended  High School or Below        2011-01-19         Employed   
9252   Extended  High School or Below        2011-02-22         Employed   
10509     Basic               College        2011-01-15         Employed   

      gender  ...  number_of_policies     policy_type        policy  \
4292       M  ...                   3  Corporate Auto  Corporate L3   
4356       F  ...                   8   Personal Auto   Personal L3   
1511       F  ...                   1   Personal Auto   Personal L3   
1571       M  ...                   3   Personal Auto   Personal L1   
7068       F  ...                   2  Corporate Auto  Corporate L3   
9230       M  ...                   3   Personal Auto   Personal L2   
7871       M  ...                   1   Personal Auto   Personal L2   
4431       M  ...                   6   Personal Auto   Personal L3   
9252       F  ...                   1   Personal Auto   Personal L3   
10509      F  ...                   1  Corporate Auto  Corporate L1   

       renew_offer_type  sales_channel  total_claim_amount  vehicle_class  \
4292             Offer1          Agent           15.169017   Two-Door Car   
4356             Offer3    Call Center          525.600000  Four-Door Car   
1511             Offer1          Agent          307.200000  Four-Door Car   
1571             Offer3            Web          494.400000     Sports Car   
7068             Offer3          Agent          161.419528  Four-Door Car   
9230             Offer4          Agent          297.600000  Four-Door Car   
7871             Offer1          Agent          326.787526            SUV   
4431             Offer4    Call Center          628.800000     Sports Car   
9252             Offer2          Agent          277.582137  Four-Door Car   
10509            Offer4         Branch          331.200000   Two-Door Car   

       vehicle_size vehicle_type month  
4292        Medsize            A     2  
4356        Medsize            A     1  
1511        Medsize            A     1  
1571        Medsize            A     2  
7068        Medsize            A     1  
9230          Small            A     1  
7871        Medsize            A     2  
4431          Small            A     1  
9252        Medsize            A     2  
10509         Small            A     1  

[10 rows x 27 columns]

In [136]:
pd.pivot_table(df, values='customer_lifetime_value', index='sales_channel', aggfunc='sum').round(2)

customer_lifetime_value
sales_channel                         
Agent                      33057887.85
Branch                     24359201.21
Call Center                17364288.37
Web                        12697632.90

#Agent: This sales channel generates the highest total customer lifetime value of $33,057,887.85, indicating its high effectiveness.

#Branch: Comes second with a total customer lifetime value of $24,359,201.21, making it a significant revenue source.

#Call Center: Ranks third with a total customer lifetime value of $17,364,288.37, generating less revenue than agents and branches.

#Web: Least effective sales channel with a total customer lifetime value of $12,697,632.90, suggesting improvement opportunities.

In [137]:
pd.pivot_table(df, values='customer_lifetime_value', index='gender', columns='education', aggfunc='mean').round(2)

education  Bachelor  College   Doctor  High School or Below   Master
gender                                                              
F           7874.27  7748.82  7328.51               8675.22  8157.05
M           7703.60  8052.46  7415.33               8149.69  8168.83

# Overall, both men and women with a Bachelor's or Master's degree have similar average customer lifetime values.
# 
# Holders of a Doctorate degree, regardless of gender, have the lowest average customer lifetime value

1. You work at the marketing department and you want to know which sales channel brought the most sales in terms of total revenue. Using pivot, create a summary table showing the total revenue for each sales channel (branch, call center, web, and mail).
Round the total revenue to 2 decimal points.  Analyze the resulting table to draw insights.

2. Create a pivot table that shows the average customer lifetime value per gender and education level. Analyze the resulting table to draw insights.

## Bonus

You work at the customer service department and you want to know which months had the highest number of complaints by policy type category. Create a summary table showing the number of complaints by policy type and month.
Show it in a long format table.

*In data analysis, a long format table is a way of structuring data in which each observation or measurement is stored in a separate row of the table. The key characteristic of a long format table is that each column represents a single variable, and each row represents a single observation of that variable.*

*More information about long and wide format tables here: https://www.statology.org/long-vs-wide-data/*

In [140]:
# Your code goes here
df['number_of_open_complaints'] = df['number_of_open_complaints'].astype(int)

# Group by 'policy_type' and 'month', and count the number of complaints
summary_table = df.groupby(['policy_type', 'month'])['number_of_open_complaints'].sum().reset_index()

# Display the summary table in long format
summary_table

policy_type  month  number_of_open_complaints
0  Corporate Auto      1                        415
1  Corporate Auto      2                        361
2   Personal Auto      1                       1635
3   Personal Auto      2                       1363
4    Special Auto      1                         84
5    Special Auto      2                         91